### ISSUE TITLE
SSH encrypted backup hangs indefinitely

### ISSUE THREAD

**Issue by liwo** [bit-team/backintime#472]
Hi,

I try to setup Back in Time with ssh encrypted backup on a synology diskstation. So far I could not get it to work on Kubuntu 15.04 with Back in Time from stable ppa version 1.1.8. python3-keyring and python3-secretstorage are both installed.

When I start `backintime-qt4 --debug` it always hangs at the command
```
DEBUG: [common/encfstools.py:93 EncFS_mount._mount] Call mount command: encfs --extpass=backintime-askpass --reverse / /home/liwo/.local/share/backintime/mnt/E5CC4CC8/mountpoint
```

If I strace this `encfs` process, I get
```
Process 24237 attached
read(0, 
```
It hangs there until I detach the process again.

Things I noticed:
When starting without any prior configuration, I get
```
DEBUG: [common/mount.py:50 Mount.__init__] pw-cache is not running
DEBUG: [common/mount.py:58 Mount.__init__] Call command: backintime pw-cache start
ERROR: Back In Time is not configured!
ERROR: [common/mount.py:64 Mount.__init__] Failed to start pw-cache: Command '['backintime', 'pw-cache', 'start']' returned non-zero exit status 2
DEBUG: [common/sshtools.py:192 SSH.unlock_ssh_agent] Add private key /home/liwo/.ssh/XXX to ssh agent
DEBUG: [common/sshtools.py:197 SSH.unlock_ssh_agent] Password available: True
DEBUG: [common/sshtools.py:213 SSH.unlock_ssh_agent] Provide password through temp FIFO
DEBUG: [common/sshtools.py:374 SSH.check_ping_host] Check ping host
DEBUG: [common/sshtools.py:383 SSH.check_ping_host] Host xxx is available
DEBUG: [common/sshtools.py:248 SSH.check_fuse] Check fuse

/* Remote probing redacted */

DEBUG: [common/sshtools.py:526 SSH.check_remote_commands] Command returncode: 0
DEBUG: [common/encfstools.py:178 EncFS_mount.check_fuse] Check fuse
DEBUG: [common/encfstools.py:197 EncFS_mount.check_version] Check version
DEBUG: [common/encfstools.py:178 EncFS_SSH.check_fuse] Check fuse
DEBUG: [common/encfstools.py:197 EncFS_SSH.check_version] Check version
DEBUG: [common/sshtools.py:244 SSH.unlock_ssh_agent] Private key /home/liwo/.ssh/XXX is already unlocked in ssh agent
DEBUG: [common/encfstools.py:258 EncFS_SSH.mount] Mount sshfs
DEBUG: [common/mount.py:298 SSH.mountprocess_lock_acquire] Acquire mountprocess lock /home/liwo/.local/share/backintime/mnt/24382.lock
DEBUG: [common/sshtools.py:124 SSH._mount] Call mount command: sshfs -p 22 -o ServerAliveInterval=240 -o IdentityFile=/home/liwo/.ssh/backup -o idmap=user liwo-backup@xxx:/volume1/Backup-XXX /home/liwo/.local/share/backintime/mnt/1CF56EBE/mountpoint
INFO: [common/mount.py:211 SSH.mount] mount ssh_encfs: liwo-backup@xxx:/volume1/Backup-XXX on /home/liwo/.local/share/backintime/mnt/1CF56EBE/mountpoint
DEBUG: [common/mount.py:319 SSH.set_mount_lock] Set mount lock /home/liwo/.local/share/backintime/mnt/1CF56EBE/locks/24382.tmp.lock
DEBUG: [common/mount.py:307 SSH.mountprocess_lock_release] Release mountprocess lock /home/liwo/.local/share/backintime/mnt/24382.lock
DEBUG: [common/encfstools.py:260 EncFS_SSH.mount] Mount local filesystem root with encfs --reverse
DEBUG: [common/mount.py:298 EncFS_mount.mountprocess_lock_acquire] Acquire mountprocess lock /home/liwo/.local/share/backintime/mnt/24382.lock
DEBUG: [common/encfstools.py:157 EncFS_mount.is_configured] Found encfs config in /home/liwo/.local/share/backintime/mnt/1CF56EBE/mountpoint/.encfs6.xml
DEBUG: [common/encfstools.py:93 EncFS_mount._mount] Call mount command: encfs --extpass=backintime-askpass --reverse / /home/liwo/.local/share/backintime/mnt/E5CC4CC8/mountpoint
```

If I try to start pw-cache manually, I get:
```
$ backintime --debug pw-cache start
DEBUG: [common/backintime.py:450 arg_parse] Arguments: {'func': <function pwCache at 0x7f4ff5687d90>, 'ACTION': 'start', 'debug': True, 'command': 'pw-cache'} | unknownArgs: []

Back In Time
Version: 1.1.8

Back In Time comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions; type `backintime --license' for details.

DEBUG: [common/backintime.py:496 getConfig] config file: /home/liwo/.config/backintime
DEBUG: [common/backintime.py:497 getConfig] profiles: ['1']
ERROR: [common/backintime.py:507 getConfig] Back In Time is not configured!
```

If I create a local profile first I can still not start pw-cache:
```
$ backintime --debug pw-cache start
DEBUG: [common/backintime.py:450 arg_parse] Arguments: {'command': 'pw-cache', 'func': <function pwCache at 0x7fbefe498d90>, 'debug': True, 'ACTION': 'start'} | unknownArgs: []

Back In Time
Version: 1.1.8

Back In Time comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions; type `backintime --license' for details.

DEBUG: [common/backintime.py:496 getConfig] config file: /home/liwo/.config/backintime
DEBUG: [common/backintime.py:497 getConfig] profiles: ['1']
DEBUG: [common/password.py:57 Password_Cache.daemonize] start
DEBUG: [common/password.py:93 Password_Cache.daemonize] write pidfile
DEBUG: [common/password.py:264 Password_Cache.run] Keyring supported: True
DEBUG: [common/password.py:269 Password_Cache.run] Nothing to cache. Quit.
DEBUG: [common/password.py:107 Password_Cache.delpid] clean-up pid file
```

The output from changing the profile to ssh encrypted doesn't change much:
```
DEBUG: [common/mount.py:50 Mount.__init__] pw-cache is not running
DEBUG: [common/mount.py:58 Mount.__init__] Call command: backintime pw-cache start
DEBUG: [common/sshtools.py:192 SSH.unlock_ssh_agent] Add private key /home/liwo/.ssh/XXX to ssh agent
DEBUG: [common/sshtools.py:197 SSH.unlock_ssh_agent] Password available: True
DEBUG: [common/sshtools.py:213 SSH.unlock_ssh_agent] Provide password through temp FIFO
DEBUG: [common/sshtools.py:374 SSH.check_ping_host] Check ping host
DEBUG: [common/sshtools.py:383 SSH.check_ping_host] Host xxx is available
DEBUG: [common/sshtools.py:248 SSH.check_fuse] Check fuse

/* Remote probing redacted */

DEBUG: [common/sshtools.py:526 SSH.check_remote_commands] Command returncode: 0
DEBUG: [common/encfstools.py:178 EncFS_mount.check_fuse] Check fuse
DEBUG: [common/encfstools.py:197 EncFS_mount.check_version] Check version
DEBUG: [common/encfstools.py:178 EncFS_SSH.check_fuse] Check fuse
DEBUG: [common/encfstools.py:197 EncFS_SSH.check_version] Check version
DEBUG: [common/sshtools.py:244 SSH.unlock_ssh_agent] Private key /home/liwo/.ssh/XXX is already unlocked in ssh agent
DEBUG: [common/encfstools.py:258 EncFS_SSH.mount] Mount sshfs
DEBUG: [common/mount.py:298 SSH.mountprocess_lock_acquire] Acquire mountprocess lock /home/liwo/.local/share/backintime/mnt/24545.lock
DEBUG: [common/sshtools.py:124 SSH._mount] Call mount command: sshfs -p 22 -o ServerAliveInterval=240 -o IdentityFile=/home/liwo/.ssh/XXX -o idmap=user liwo-backup@xxx:/volume1/Backup-XXX /home/liwo/.local/share/backintime/mnt/1CF56EBE/mountpoint
INFO: [common/mount.py:211 SSH.mount] mount ssh_encfs: liwo-backup@xxx:/volume1/Backup-XXX on /home/liwo/.local/share/backintime/mnt/1CF56EBE/mountpoint
DEBUG: [common/mount.py:319 SSH.set_mount_lock] Set mount lock /home/liwo/.local/share/backintime/mnt/1CF56EBE/locks/24545.tmp.lock
DEBUG: [common/mount.py:307 SSH.mountprocess_lock_release] Release mountprocess lock /home/liwo/.local/share/backintime/mnt/24545.lock
DEBUG: [common/encfstools.py:260 EncFS_SSH.mount] Mount local filesystem root with encfs --reverse
DEBUG: [common/mount.py:298 EncFS_mount.mountprocess_lock_acquire] Acquire mountprocess lock /home/liwo/.local/share/backintime/mnt/24545.lock
DEBUG: [common/encfstools.py:157 EncFS_mount.is_configured] Found encfs config in /home/liwo/.local/share/backintime/mnt/1CF56EBE/mountpoint/.encfs6.xml
DEBUG: [common/encfstools.py:93 EncFS_mount._mount] Call mount command: encfs --extpass=backintime-askpass --reverse / /home/liwo/.local/share/backintime/mnt/E5CC4CC8/mountpoint
```

Starting with an encrypted local profile and changing that to ssh encrypted doesn't help either other than pw-cache finally starts. But encfs hangs nontheless.

Sorry for the long ticket, I'm not sure which part of this information is relevant and which isn't.

**Comment by Germar** [comment]
Just to make sure: did you follow this [HowTo for Synology NAS](http://wiki.qedinux.be/index.php?title=BackInTime_to_Synology_DSM_5)?

Long tickets are fine :+1:
pw-cache behavior is fine. Nothing to worry about.

I'll test it in a Kubuntu 15.04 VM after dinner...

**Comment by Germar** [comment]
It does work here with Kubuntu 15.04. Can you please try again with a fresh remote path (which has no `.encfs6.xml`)?

**Comment by liwo** [comment]
Hi Germar,

thanks for your feedback. I followed the steps in this HowTo, maybe not exactly that one but at least a similar one. I don't think the remote side is a problem here because in the command that hangs no remote resource is involved as far as I can tell (`encfs --reverse / /mountpoint`).

I tried again with an empty remote path and I removed `rm ~/.config/backintime/config` and `rm -r ~/.local/share/backintime/*` so hopefully started fresh over. The result is the same, it hangs at
```
DEBUG: [common/encfstools.py:93 EncFS_mount._mount] Call mount command: encfs --extpass=backintime-askpass --reverse / /home/liwo/.local/share/backintime/mnt/E5CC4CC8/mountpoint
```

I think it might be a problem with backintime-askpass not providing the password and encfs is waiting to read it, but that is just speculation on my part (backed by the fact that `strace` shows that encfs is waiting for a `read()` to finish). 

Can I do anything to debug this further? I have never programmed in python, so not sure about the ecosystem with debugging and stuff, but if you can give me a pointer in which direction I could start I could give that a try.

**Comment by Germar** [comment]
The `backintime-askpass` process reads from a temp FIFO spawned by the main backintime process in this case. So `pw-cache` is not involved here. I just missed to add that debug message :confounded:

There where similar bugs with de-/encode by `encfsctl` before. Which where cause by wrong buffer size. Please download [this patch](https://gist.github.com/Germar/2586315e0849c7563c2f#file-encfstools-patch) and apply it with:

    sudo patch /usr/share/backintime/common/encfstools.py < encfstools.patch

**Comment by liwo** [comment]
The patch didn't help unfortunately. I went over everything again and found one interesting line in the log:
```
DEBUG: [common/encfstools.py:158 EncFS_mount.is_configured] Found encfs config in /home/liwo/.local/share/backintime/mnt/1CF56EBE/mountpoint/.encfs6.xml
```
Which shouldn't be there because I deleted this file before. Turns out I never realized that the bind mount from the HowTo was not mounted anymore for some reason I have yet to figure out. And the premissions in the directory /volume1/volume1/Backup-XXX where weird, basically `chmod 0`. Which explains why things break but not really why encfs hangs on read. After all, I get a permission denied if I try to `cat` the file. 

If I mount the directory correctly and remove everything again I get farther, now the creation of the config file fails, but that seems to be a permission problem on the Synology. It's late tonight, that problem remains for another evening.

Sorry for bothering you with this as it probably was some error on my part (or some Synology weirdness?) and thanks for your efforts and time!

**Comment by liwo** [comment]
Ok, I got it to work now. In the end it turns out that the problem I saw actually was the effect of some weird synology settings: In (guessing english labels by translating from german) "Shared Folders" > Folder Settings > Advanced the checkboxes for "disallow via File Station, FTP and WebDAV" were all set (disallow listing of files, disallow changing of files, disallow download of files). Obviosly these apply to SFTP / SSHFS as well. Took me a while to figure that out, as it works fine when connecting via SSH to the box but not via sftp. 

So thanks again for pointing me in the right direction!
